main.go

```diff
@@ -17,90 +20,102 @@ var (
 	magefile string
 	//go:embed example/Makefile
 	makefile string
+	//go:embed example/.mage/mgmake_gen.go
+	mgmake string
+	// nolint: gochecknoglobals
+	mageDir = mgpath.FromGitRoot(mgpath.MageDir)
 )
 
 func main() {
+	logger := mglog.Logger("mage-tools")
 	usage := func() {
-		logger := mglog.Logger("main")
 		logger.Info(`Usage:
 	init	to initialize mage-tools`)
 		os.Exit(0)
 	}
 	if len(os.Args) <= 1 {
 		usage()
 	}
-
 	switch os.Args[1] {
 	case "init":
-		initMageTools()
+		if err := initMageTools(); err != nil {
+			log.Fatalf(err.Error())
+		}
+	case "gen":
+		if err := gen(); err != nil {
+			log.Fatalf(err.Error())
+		}
 	default:
 		usage()
 	}
 }
 
-func initMageTools() {
+func gen() error {
+	mglog.Logger("gen").Info("generating makefiles...")
+	executable := filepath.Join(mgpath.Tools(), "mgmake", "magefile")
+	if err := mgtool.RunInDir("git", mageDir, "clean", "-fdx", filepath.Dir(executable)); err != nil {
+		return err
+	}
+	if err := os.WriteFile(filepath.Join(mageDir, mgpath.MakeGenGo), []byte(mgmake), 0o600); err != nil {
+		return err
+	}
+	if err := mgtool.RunInDir("go", mageDir, "mod", "tidy"); err != nil {
+		return err
+	}
+	if exit := mage.ParseAndRun(os.Stdout, os.Stderr, os.Stdin, []string{"-compile", executable}); exit != 0 {
+		return fmt.Errorf("faild to compile magefile binary")
+	}
+	return mgtool.RunInDir(executable, mageDir, mgpath.GenMakefilesTarget, executable)
+}
+
+func initMageTools() error {
 	logger := mglog.Logger("init")
 	logger.Info("initializing mage-tools...")
-	defer func() {
-		if err := recover(); err != nil {
-			logger.Error(err.(error), err.(error).Error())
-		}
-	}()
 
-	if mgtool.GetCWDPath("") != mgtool.GetGitRootPath("") {
-		panic(fmt.Errorf("can only be generated in git root directory"))
+	if mgpath.FromWorkDir("") != mgpath.FromGitRoot("") {
```

Comment:
would there be any difference between "." and "" here?

Reply:
"." would be more clear IMO

Reply:
No difference, both works